src/coreclr/jit/emitriscv64.cpp

```diff
@@ -4977,73 +4979,72 @@ regNumber emitter::emitInsTernary(instruction ins, emitAttr attr, GenTree* dst,
-                        emitIns_R_R_I(INS_srli, attr, tempReg2, saveOperReg2, ui6);
-
-                        emitIns_R_R_R(INS_xor, attr, tempReg1, tempReg1, tempReg2);
-                        if (attr == EA_4BYTE)
-                        {
-                            imm = 1;
-                            emitIns_R_R_I(INS_andi, attr, tempReg1, tempReg1, imm);
-                            emitIns_R_R_I(INS_andi, attr, tempReg2, tempReg2, imm);
-                        }
-                        // if (B > 0 && C < 0) || (B < 0  && C > 0), skip overflow
-                        BasicBlock* tmpLabel  = codeGen->genCreateTempLabel();
-                        BasicBlock* tmpLabel2 = codeGen->genCreateTempLabel();
-                        BasicBlock* tmpLabel3 = codeGen->genCreateTempLabel();
+                        resultReg = saveOperReg1;
+                        regOp1    = dstReg;
+                        regOp2    = saveOperReg2;
+                    }
 
-                        emitIns_J_cond_la(INS_bne, tmpLabel, tempReg1, REG_R0);
+                    instruction branchIns  = INS_none;
+                    regNumber   branchReg1 = REG_NA;
+                    regNumber   branchReg2 = REG_NA;
 
-                        emitIns_J_cond_la(INS_bne, tmpLabel3, tempReg2, REG_R0);
+                    if (((dst->gtFlags & GTF_UNSIGNED) != 0))
+                    {
+                        // if B > A then overflow
+                        branchIns  = INS_bltu;
+                        branchReg1 = resultReg;
+                        branchReg2 = regOp1;
+                    }
+                    else
+                    {
+                        regNumber tempReg1 = dst->ExtractTempReg();
 
-                        // B > 0 and C > 0, if A < B, goto overflow
-                        emitIns_J_cond_la(INS_bge, tmpLabel, isAdd ? dstReg : saveOperReg1,
-                                          isAdd ? saveOperReg1 : saveOperReg2);
+                        branchIns = INS_bne;
 
-                        codeGen->genDefineTempLabel(tmpLabel2);
+                        if ((ins == INS_addw) && (attr == EA_8BYTE))
+                        {
+                            assert((src1->gtType == TYP_INT) && (src2->gtType == TYP_INT));
 
-                        codeGen->genJumpToThrowHlpBlk(EJ_jmp, SCK_OVERFLOW);
+                            emitIns_R_R_R(INS_add, attr, tempReg1, regOp1, regOp2);
 
-                        codeGen->genDefineTempLabel(tmpLabel3);
+                            // if 64-bit addition is not equal to 32-bit addition for 32-bit operands then overflow
+                            branchReg1 = resultReg;
+                            branchReg2 = tempReg1;
+                        }
+                        else
+                        {
+                            assert(tempReg != tempReg1);
+                            regNumber tempReg2 = tempReg; // When tempReg2 is used the tempReg has to be already dead
```

Comment:
Then you should probably remove one buildInternalIntRegisterDefForNode call from LinearScan::BuildNode.